docs/event-serializing.md

```diff
@@ -0,0 +1,139 @@
+---
+id: event-serializing
+title: Event Serializing
+sidebar_label: Event Serializing
+---
+
+We support [Apache Avro](https://avro.apache.org/) and standard JSON (de)serialization of events for producer as well as for consumer. This is working together with [Cloud Events specification](https://github.com/cloudevents/spec) (versions `0.1` and `0.2`). To make both of these worlds work we are utilizing Kafka headers introduced in version `0.11`. As Cloud Events specify there are 2 modes how to make this combination work. Discussion about this topic can be followed on [Github](https://github.com/cloudevents/spec/pull/337/files).
+
+**Structured content mode**
+
+This mode should be used when JSON format is desired, thus no further serialization and deserialization such as Apache Avro.
+
+> ##### KAFKA EVENT HEADERS
+>
+> ```
+> cloudEvent_contentType: "application/json"
+> ```
+>
+> ##### KAFKA EVENT VALUE
+>
+> ```
+> {
+>    "specversion" : "0.1",
+>    "type" : "com.example.someevent",
+>    "data": {},
+>    ...
+> }
+> ```
+
+**Binary content mode**
+
+This mode should be used when further serialization and deserialization is desired (e.g. Apache Avro).
+
+> ##### KAFKA EVENT HEADERS
+>
+> ```
+> cloudEvent_contentType: "avro/binary"
+> cloudEvent_specversion: "0.2"
+> cloudEvent_type: "com.example.someevent"
+> ...
+> ```
+>
+> ##### KAFKA EVENT VALUE
+>
+> ```
+> ... application data ...
+> ```
+
+## Apache Avro format
+
+To be compatible across the board Apache Avro is using specific [format](https://docs.confluent.io/current/schema-registry/docs/serializer-formatter.html#wire-format) that needs to be followed.
+
+```bash
+# 0 - magic byte
+# 1-4 - schema id - this is used by consumer to know which schema to use for deserialization
+# 5-... - data
+
+# Example
+<0, 0, 0, 0, 1, 5, 3, 8, ...>
+```
+
+### Setup
+
+![event-serializing-avro](assets/event-serializing-avro.png)
+
+Setup for event (de)serialization if fairly straightforward. Besides classic components like web client, RIG and Kafka it needs also `Kafka Schema Registry`. `Kafka Schema Registry` is place where all schemas are stored and retrieved from. As event is consumed from Kafka, RIG fetches schema, caches it in-memory and does deserialization. After that event is sent to web client (or any other client able to connect via WS/SSE). As for producing, RIG fetches schema, caches it in-memory and serialize event. Then it's sent to Kafka.
+
+### Producer details
+
+- producer evaluates if serializing is turned on by checking `KAFKA_SERIALIZER` environment variable and if it's value is `avro`
+- If it is, creates headers for Kafka event by appending `cloudEvents_` prefix for every field besides `data` field
+  - **for deep nested values we are using query encoding** since Kafka headers don't support nested values
```

Comment:
perhaps you should point out that this only applies to so-called CloudEvents "extensions"

Reply:
Do we want to try stringified values instead of this monster?